File: ui/src/features/agents/lib/thinkTags.ts
/**
 * Split leaked reasoning-tag blocks out of model text. Some open models served
 * through OpenAI-compatible gateways emit reasoning inline as tagged blocks
 * (`<think>` for DeepSeek/Qwen/Kimi, `<mm:think>` for MiniMax-M3) instead of a
 * separate reasoning field. MiniMax also emits a bare closing tag on
 * non-thinking turns, and a stream can end mid-thought with an unclosed
 * opener — text on the thinking side of either is treated as reasoning.
 */
export type ThinkSegment = { kind: "reasoning" | "text"; text: string };

const MARKER = /<(\/?)(?:think|thinking|mm:think)>/gi;

export function splitThinkTags(text: string): Array<ThinkSegment> {
  const segments: Array<ThinkSegment> = [];
  let mode: ThinkSegment["kind"] = "text";
  let last = 0;

  const push = (kind: ThinkSegment["kind"], end: number) => {
    if (end > last) segments.push({ kind, text: text.slice(last, end) });
  };

  for (const match of text.matchAll(MARKER)) {
    const isCloser = match[1] === "/";
    const index = match.index ?? 0;
    if (isCloser) {
      // A closer without an opener (MiniMax non-thinking turns) still means
      // everything before it was reasoning.
      push("reasoning", index);
      mode = "text";
      last = index + match[0].length;
    } else if (mode === "text") {
      push("text", index);
      mode = "reasoning";
      last = index + match[0].length;
    }
    // An opener while already inside reasoning is left as part of the block.
  }

  push(mode, text.length);
  return segments;
}

File: agent/utils/think_tags.py
"""Strip leaked reasoning-tag blocks from model text.

Some open models served through OpenAI-compatible gateways leak reasoning into
visible text as tagged blocks (`<think>` for DeepSeek/Qwen/Kimi, `<mm:think>`
for MiniMax-M3) instead of a separate reasoning field. MiniMax also emits a
bare closing tag on non-thinking turns and can end a stream mid-thought with
an unclosed opener.

Only use the stripped text to decide whether a message has user-visible
content — never write it back into the message, since MiniMax requires
thinking blocks to be passed back verbatim in conversation history.
"""

import re

_TAG = r"(?:think|thinking|mm:think)"
_PAIRED = re.compile(rf"<({_TAG})>.*?</\1>", re.DOTALL | re.IGNORECASE)
_BARE_CLOSER = re.compile(rf"\A.*</{_TAG}>", re.DOTALL | re.IGNORECASE)
_UNCLOSED_OPENER = re.compile(rf"<{_TAG}>.*\Z", re.DOTALL | re.IGNORECASE)


def strip_think_tags(text: str) -> str:
    text = _PAIRED.sub("", text)
    text = _BARE_CLOSER.sub("", text)
    text = _UNCLOSED_OPENER.sub("", text)
    return text
